<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Заполнение-пропусков." data-toc-modified-id="Заполнение-пропусков.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Заполнение пропусков.</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-2.8.1"><span class="toc-item-num">2.8.1&nbsp;&nbsp;</span>Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод--по-вопросу-1:" data-toc-modified-id="Вывод--по-вопросу-1:-2.8.2"><span class="toc-item-num">2.8.2&nbsp;&nbsp;</span>Вывод  по вопросу 1:</a></span></li><li><span><a href="#Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-2.8.3"><span class="toc-item-num">2.8.3&nbsp;&nbsp;</span>Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-по-вопросу-2:" data-toc-modified-id="Вывод-по-вопросу-2:-2.8.4"><span class="toc-item-num">2.8.4&nbsp;&nbsp;</span>Вывод по вопросу 2:</a></span></li><li><span><a href="#Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-2.8.5"><span class="toc-item-num">2.8.5&nbsp;&nbsp;</span>Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-по-вопросу-3:" data-toc-modified-id="Вывод-по-вопросу-3:-2.8.6"><span class="toc-item-num">2.8.6&nbsp;&nbsp;</span>Вывод по вопросу 3:</a></span></li><li><span><a href="#Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-2.8.7"><span class="toc-item-num">2.8.7&nbsp;&nbsp;</span>Вопрос 4: Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Вывод-по-вопросу-4:" data-toc-modified-id="Вывод-по-вопросу-4:-2.8.8"><span class="toc-item-num">2.8.8&nbsp;&nbsp;</span>Вывод по вопросу 4:</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Исследование надежности заемщиков.

По предоставленным данным кредитного отдела банка в виде статистики о платёжеспособности клиентов, необходимо провести исследование относительно влияния ряда факторов (например: семейного положения и количества детей клиента) на факт погашения кредита в срок.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Обзор данных

In [1]:
# импортирование необходимых библиотек
import numpy as np 
import pandas as pd

In [2]:
# чтение файла с данными и сохранение в df
df = pd.read_csv('data.csv')

In [3]:
# получение первых 10 строк таблицы df 
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [4]:
# получение общей информации о данных
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
# получение статистики по каждому числовому признаку
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

**Выводы**

Согласно документации к данным в таблице 12 столбцов и 21525 строк. 

В документации имеется следующее описание столбцов:
* `children` — количество детей в семье;
* `days_employed` — общий трудовой стаж в днях;  
* `dob_years` — возраст клиента в годах;
* `education` — уровень образования клиента;
* `education_id` — идентификатор уровня образования;
* `family_status` — семейное положение;
* `family_status_id` — идентификатор семейного положения;
* `gender` — пол клиента;
* `income_type` — тип занятости;
* `debt` — имел ли задолженность по возврату кредитов;
* `total_income` — ежемесячный доход;
* `purpose` — цель получения кредита.

Исходя из описания поставленной задачи, предоставленные документы являются статистикой о платёжеспособности клиентов одного из кредитных учреждений. Заголовки столбцов указаны корректно, к каждому из столбцов имеется описание содержащихся в них данных.

Предварительно можно утверждать, что, размера предоставленных данных достаточно для проверки гипотез, однако начальное ознакомление с данными позволило выявить следующие проблемы:
- пропущенные значения;
- дубликаты — явные (возможно имеются неявные);
- некорректные значения в столбцах (например: отрицательные значения в размере трудового стажа);
- типы данных в некоторых столбцах некорректны применительно к находящимся в них данных.

## Предобработка данных

### Заполнение пропусков.

In [6]:
# подсчёт пропусков
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [7]:
# выяснение процентного соотношения пропусков в столбцах к общему числу значений в столбце
df.isna().mean()

children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64

В предоставленных данных встречаются пропуски, а именно: в столбцах с общим трудовым стажем (days_employed) 
и ежемесячным доходом (total_income). Пропущенные значения составляют около 10 % от общего числа количественных значений в каждом из столбцов, что может негативно отразиться на результате проверяемых гипотез.

Причинами появления пропусков в указанных столбцах могут быть следующими:
* условия предоставления денежных средств по некоторым кредитным продуктам допускают отсутствие достоверных сведений о доходе и трудовом стаже (т.е. без предоставления трудовой книжки и справки о доходе), в связи с чем соответствующие данные заполняются со слов потенциальных заемщиков. Часть заемщиков может иметь заработок и стаж без официального оформления на работу, что во внутрибанковской системе может отражаться как отсутствующие значения;
* стаж и размер заработка могут быть не отражены, поскольку потенциальный клиент только приступил к трудовой деятельности и ранее дохода не получал.

Замена пропущенных значений медианой является приемлемым, поскольку медиана не так чувствительна к выбросам по сравнению со средним арифметическим значением, что позволит не сильно исказить значения в соответствующих столбцах.

Как мы видим, медианное значение в столбце days_employed (стаж работы) является отрицательным, что в реальности быть не может, кроме того, т.е. в указанном столбце ряд значений имеют признак аномалий. В связи с чем, до замены пропусков в соответствующем столбце необходимо будет избавиться от всех имеющихся аномалий.

Посмотрим имеются ли аномалии в столбце с данными о ежемесячной заработной плате

In [8]:
# посчитаем медиану значений столбца с данными о ежемесячной заработной плате, округлив результат до целого числа
df.total_income.median().round()

145018.0

Медианный размер ежемесячного дохода довольно велик с учетом того, что среди заемщиков имеются пенсионеры, работники в декрете и студенты.
Посмотрим, насколько сильно отклоняется данное значение при группировке по каждому типу занятости.

In [9]:
# сгруппируем данные по медианному доходу по типу занятости
df.groupby('income_type')['total_income'].median().round()

income_type
безработный        131340.0
в декрете           53829.0
госслужащий        150448.0
компаньон          172358.0
пенсионер          118514.0
предприниматель    499163.0
сотрудник          142594.0
студент             98202.0
Name: total_income, dtype: float64

Размер ежемесячной заработной платы значительно отличается в зависимости от типа занятости, в связи с чем подставление во все столбцы одинакового медианного значения безотносительно типа занятости может привести к некорретным результатам исследования.

In [10]:
# для каждой строки со значением из столбца с типом занятости и при отсутствующем значении в столбце с total_income 
# подставим медианное значение заработной платы в столбец total_income 
for value in df['income_type'].unique():
    try: df.loc[(df['income_type'] == value) & (df['total_income'].isna()), 'total_income'] \
        = df.loc[df['income_type'] == value, 'total_income'].median()
    except:
        # выведем сообщение об ошибке
        # выведем какие типы данных имеются в самом столбце и посмотрим на отсутствие типа str в заменяемых значениях
        print('Error')
        print(df['total_income'].apply(type).value_counts())      

In [11]:
# проверим отсутствие пропусков в столбце total_income после преобразований
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

### Проверка данных на аномалии и исправления.

Повторно взглянем на статистику по каждому числовому признаку

In [12]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  2.152500e+04  
mean           0.972544      0.080883  1.652253e+05  
std            1.420324      0.272661  9.804367e+04  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.077982e+05  
50%            0.000000      0.000000  1.425944e+05  
75%            1.000000      0.000000  1.955499e+05  
max            4.000000      1.000000  2.265604e+06

В предоставленных данных имеются следующие аномалии:
- в столбце `children` минимальное значение является не "0", а "-1"
- в столбце `days_employed` имеются отрицательные значения, а максимальный стаж и среднее арифметическое кратно превышают срок человеческой жизни

In [13]:
# посмотрим на значения, имеющиеся в столбце с данными о детях
df.children.value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

В столбце с данными о детях, помимо отрицательного значения, выявилась еще одна аномалия: большое количество клиентов, у которых указано 20 детей, что маловероятно. Вероятно случайно был добавлен лишний ноль в данных во время выгрузки из базы или ввиду некорректного внесения информации сотрудниками, заменим данные значения на медианные

In [14]:
# скорректируем отрицательное значение и выброс на медианные значения 
df.loc[(df['children'] == -1) | (df['children'] == 20), 'children'] = df.children.median()

In [15]:
# проверим корректность проведенных преобразований
df.children.value_counts()

0.0    14272
1.0     4818
2.0     2055
3.0      330
4.0       41
5.0        9
Name: children, dtype: int64

Теперь подробнее рассмотрим столбец с данными о трудовом стаже

In [16]:
# посмотрим какое количество данных в столбце имеют отрицательные значения 
df.loc[df['days_employed'] < 0]

children  days_employed  dob_years       education  education_id  \
0           1.0   -8437.673028         42          высшее             0   
1           1.0   -4024.803754         36         среднее             1   
2           0.0   -5623.422610         33         Среднее             1   
3           3.0   -4124.747207         32         среднее             1   
5           0.0    -926.185831         27          высшее             0   
...         ...            ...        ...             ...           ...   
21519       1.0   -2351.431934         37  ученая степень             4   
21520       1.0   -4529.316663         43         среднее             1   
21522       1.0   -2113.346888         38         среднее             1   
21523       3.0   -3112.481705         38         среднее             1   
21524       2.0   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
5      гражданский брак                 1      M   компаньон     0   
...                 ...               ...    ...         ...   ...   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
0      253875.639453                      покупка жилья  
1      112080.014102            приобретение автомобиля  
2      145885.952297                      покупка жилья  
3      267628.550329         дополнительное образование  
5      255763.565419                      покупка жилья  
...              ...                                ...  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля  

[15906 rows x 12 columns]

Превалирующее число значений (более 70%) в столбце `days_employed` имеют отрицательное значение, что вероятно произошло вследствие ошибки внесения соответствующих данных или переноса их из другой системы. Исключение такого значительного объема информации критично для корректности результатов нашего исследования. Разумным вариантом решения данной проблемы будет приведение всех значений в столбце к положительным значениям.

In [17]:
# преобразуем отрицательные значения в столбце со стажем в положительные
df['days_employed'] = df['days_employed'].abs()

In [18]:
# проверим корректность произведенных преобразований и отсутствие отрицательных значений в столбце
df[df['days_employed'] < 0]

Empty DataFrame
Columns: [children, days_employed, dob_years, education, education_id, family_status, family_status_id, gender, income_type, debt, total_income, purpose]
Index: []

In [19]:
# посчитаем медиану значений данного столбца после преобразований, округлив результат до целого числа
df.days_employed.median().round()

2194.0

Медианный размер трудового стажа в днях, эквивалентный 6 годам, вполне может соответствовать реальным значениям, заменим им пропущенные значения в столбце `days_employed`.

In [20]:
df['days_employed'] = df['days_employed'].fillna(2169.0)

Выясним какое количество данных в столбце со стажем (`days_employed`) имеют значения, превышающие значение в столбце 
возраст заемщика (`dob_years`) минус 16 лет (минимальный возраст для начала трудовой деятельности, в т.ч. предпринимательской)

In [21]:
# переведем значения в столбце days_employed к количеству лет и применим фильтрацию
days_employed_abnormal = (df['days_employed'] / 365 > (df['dob_years']) - 16).sum()

In [22]:
days_employed_abnormal

3642

In [23]:
# заменим аномальные значения значения в столбце с общим трудовым стажем на полученное медианное значение 
df.loc[df['days_employed'] / 365 > (df['dob_years']) - 16, 'days_employed'] = df.days_employed.median().round()

In [24]:
# проверим корректность всех проведенных преобразований
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   21525.000000  21525.000000  21525.000000   
mean       0.470476    2276.197782     43.293380      0.817236   
std        0.750534    1912.804887     12.574584      0.548138   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000    1031.303984     33.000000      1.000000   
50%        0.000000    2169.000000     42.000000      1.000000   
75%        1.000000    2470.912766     53.000000      1.000000   
max        5.000000   16119.687737     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  2.152500e+04  
mean           0.972544      0.080883  1.652253e+05  
std            1.420324      0.272661  9.804367e+04  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.077982e+05  
50%            0.000000      0.000000  1.425944e+05  
75%            1.000000      0.000000  1.955499e+05  
max            4.000000      1.000000  2.265604e+06

In [25]:
# проверим отсутствие пропущенных значений
df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных.

In [26]:
# посмотрим существующие типы данных в столбцах таблицы
df.dtypes

children            float64
days_employed       float64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

В столбцах `children`, `days_employed`, `dob_years` тип данных указан как вещественное число, в столбце `total_income` указано слишком большое количество знаков после запятой, что не соответствует реальным значениям. Преобразуем указанные столбцы к целочисленному типу.

In [27]:
try:
    df[['children', 'days_employed', 'dob_years', 'total_income']] \
    = df[['children', 'days_employed', 'dob_years', 'total_income']].astype(int)
except:
    print('Error_change_type')

In [28]:
# проверим корректность произведенных преобразований типов 
df.dtypes

children             int64
days_employed        int64
dob_years            int64
education           object
education_id         int64
family_status       object
family_status_id     int64
gender              object
income_type         object
debt                 int64
total_income         int64
purpose             object
dtype: object

### Удаление дубликатов.

Посмотрим на количество явных дубликатов в таблице 

In [29]:
df.duplicated().sum()

54

In [30]:
# удалим имеющиеся явные дубликаты
df = df.drop_duplicates().reset_index(drop=True)

Для выявления неявных дубликатов, посмотрим на все имеющиеся значения в столбцах с типом данных object

In [31]:
df.education.value_counts()

среднее                13705
высшее                  4710
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   273
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

В столбце `education` имеются данные со значениями, написанными в разном регистре (с использованием заглавных и строчных букв), что затрудняет объединение этих признаков для проведения дальнейшего исследования.

In [32]:
# преобразуем значения в столбце к нижнему регистру
df.education = df.education.str.lower()

Убедимся в корректности преобразования и посмотрим, имеются ли еще какие-либо различия в написании.

In [33]:
df.education.value_counts()

среднее                15188
высшее                  5251
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Данные в столбце `education` приведены к единому стилю, посмотрим имеются ли схожие проблемы в других столбцах типа object.

In [34]:
df.family_status.value_counts()

женат / замужем          12344
гражданский брак          4163
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

В столбце family_status имеется числовое значение 0.0, которое вероятнее всего означает отсутствие брака и внесено в данный столбец ошибочно вместо столбца family_status_id. Изменим некорректное значение на "Не женат / не замужем". 

In [35]:
df.loc[(df['family_status'] == 0.0), 'family_status'] = 'Не женат / не замужем'

In [36]:
# проверим корректность проведенных преобразований
df.family_status.value_counts()

женат / замужем          12344
гражданский брак          4163
Не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

In [37]:
df.gender.value_counts()

F      14189
M       7281
XNA        1
Name: gender, dtype: int64

Значения в столбце `gender` возникли вследствие некорректного ведения информации (возможно при копировании данных в столбец с семейным статусом). Значения, отличающиеся от 'F' и 'M', невозможно идентифицировать, что делает их бесполезными для проведения анализа, удалим их.

In [39]:
# удалим пустые значения из таблицы
df = df.dropna().reset_index(drop=True)

In [40]:
# проверим корректность проведенных преобразований
df.gender.value_counts()

F    14189
M     7281
Name: gender, dtype: int64

In [41]:
df.purpose.value_counts()

свадьба                                   793
на проведение свадьбы                     773
сыграть свадьбу                           769
операции с недвижимостью                  675
покупка коммерческой недвижимости         662
покупка жилья для сдачи                   652
операции с жильем                         652
операции с коммерческой недвижимостью     650
жилье                                     646
покупка жилья                             646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          625
покупка своего жилья                      620
покупка недвижимости                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

В столбце с данными о цели кредита большое количество неявных дубликатов (словосочетаний, имеющих схожее значение), которые вероятнее всего появились вследствие занесения каждым сотрудником по своему усмотрению, без наличия в соответствующей внутренней системе регламентированных категорий для соответствующих сведений. 

### Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

In [42]:
# создадим новую таблицу и поместим в нее значения в виде id и соответствующий ему тип образования
education_dict = df[['education','education_id']]

In [43]:
education_dict = education_dict.drop_duplicates().reset_index(drop=True)

In [44]:
education_dict

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [45]:
# создадим новую таблицу и поместим в нее значения в виде id и соответствующий ему тип образования
family_status_dict = df[['family_status','family_status_id']]

In [46]:
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)

In [47]:
family_status_dict

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

In [48]:
# удалим столбцы с исходными значениями образования и семейного положения со значением из основной таблицы, 
# оставив соответствующие им столбцы с идентификаторами 
df = df.drop(['education', 'family_status'], axis=1)

In [49]:
# проверим результат преобразования
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           2169         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

### Категоризация дохода.

Воспользуемся функцией pd.cut для определения точных границ интервалов, задав параметр bin и применим указанную функцию для создания нового столбца из значений столбца с ежемесячным доходом.

In [50]:
# создадим список категорий
cut_labels = ['E', 'D', 'C', 'B', 'A']

In [51]:
# создадим список диапазонов значений применительно к каждой категории, указав границы каждого из них
cut_bins = [0, 30000, 50000, 200000, 1000000, 10000000]

In [52]:
# создадим новый столбец с категориями и применим к нему необходимые вычисления
df['total_income_category'] = pd.cut(df['total_income'], bins=cut_bins, labels=cut_labels)

In [53]:
# проверим корректность произведенных преобразований
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           2169         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

### Категоризация целей кредита.

Как мы ранее узнали, в столбце с целями кредита имеется большое количество схожих значений. 
Для того, чтобы сгруппировать их по соответствующим категориям, выясним какие слова встречаются чаще всего и применим к ним функцию для создания нового столбца с категориями.

In [54]:
# импортируем библиотеки, необходимые для обработки текстовых данных в столбце с целями кредита
from pymystem3 import Mystem
from collections import Counter
m = Mystem()

In [55]:
# посмотрим какие слова в описании целей кредита используются чаще всего
lemma = Counter(m.lemmatize(' '.join(df.purpose)))

In [56]:
lemma

Counter({'покупка': 5899,
         ' ': 55064,
         'жилье': 4461,
         'приобретение': 461,
         'автомобиль': 4308,
         'дополнительный': 907,
         'образование': 4014,
         'сыграть': 769,
         'свадьба': 2335,
         'операция': 2604,
         'с': 2918,
         'на': 2228,
         'проведение': 773,
         'для': 1290,
         'семья': 638,
         'недвижимость': 6352,
         'коммерческий': 1312,
         'жилой': 1231,
         'строительство': 1879,
         'собственный': 635,
         'подержать': 853,
         'свой': 2231,
         'со': 627,
         'заниматься': 904,
         'сделка': 941,
         'получение': 1315,
         'высокий': 1374,
         'подержанный': 111,
         'профильный': 436,
         'сдача': 652,
         'ремонт': 607,
         '\n': 1})

In [57]:
# создадим функцию, которая принимает на вход строку из столбца нашей таблицы
# и приводит строку к списку лемманизированных слов, а затем
# находит в этом списке искомое значение для отнесения ее к соответствующей категории
def category_grouped(row):
    row = m.lemmatize(row)
    if 'жилье' in row or 'недвижимость' in row:
        return 'операции с недвижимостью'
    if 'автомобиль' in row:
        return 'операции с автомобилем'
    if 'свадьба' in row:
        return 'проведение свадьбы'
    if 'образование' in row:
        return 'получение образования'

In [58]:
# применим функцию к столбцу со сведениями о целях кредита
df['purpose_category'] = df['purpose'].apply(category_grouped)

In [59]:
# посмотрим на получившиеся категории в созданном столбце
df['purpose_category'].value_counts()

операции с недвижимостью    10813
операции с автомобилем       4308
получение образования        4014
проведение свадьбы           2335
Name: purpose_category, dtype: int64

### Ответы на вопросы.

#### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для проверки данной гипотезы создадим сводную таблицу с данными по соответствующим значениям.

In [60]:
table_children = df.pivot_table(index='children', values='debt', aggfunc={'debt' : 'sum'}) \
    .reset_index() \
    .rename(columns={'debt' : 'amount_debt_delay'})

In [61]:
# выведем ее на экран
table_children 

children  amount_debt_delay
0         0               1072
1         1                444
2         2                194
3         3                 27
4         4                  4
5         5                  0

In [62]:
table_children_full = df.groupby('children', as_index=False) \
    .agg({'gender' : 'count'}) \
    .rename(columns={'gender' : 'all_client_children'})

In [63]:
# присоединим таблицу с общим количеством клиентов по количеству детей
table_children = table_children.merge(table_children_full, how='inner', on='children')

In [64]:
# добавим столбец с процентным соотношением числа клиентов с просрочками к общему числу клиентов по наличию детей
table_children['procent_debt'] = \
(table_children['amount_debt_delay'] / table_children['all_client_children'] * 100).round(2)

In [65]:
# проверим получившуюся таблицу и сделаем выводы
table_children.sort_values(by='procent_debt', ascending=False)

children  amount_debt_delay  all_client_children  procent_debt
4         4                  4                   41          9.76
2         2                194                 2052          9.45
1         1                444                 4809          9.23
3         3                 27                  330          8.18
0         0               1072                14229          7.53
5         5                  0                    9          0.00

#### Вывод  по вопросу 1:

Зависимость между количеством детей и наличием задолженности по кредитам у клиентов имеется: при наличии детей, клиенты менее ответственно относятся к исполнению своих обязательств и чаще допускают просрочки, чем клиенты не имеющие детей.

#### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

Для проверки данной гипотезы создадим сводную таблицу с данными по соответствующим значениям.

In [66]:
table_family_status = df.pivot_table(index='family_status_id', values='debt', aggfunc={'debt' : 'sum'}) \
    .reset_index() \
    .rename(columns={'debt' : 'amount_debt_delay'})

In [67]:
# выведем ее на экран
table_family_status

family_status_id  amount_debt_delay
0                 0                931
1                 1                388
2                 2                 63
3                 3                 85
4                 4                274

In [68]:
# присоединим таблицу с расшифровками семейного статуса
table_family_status = table_family_status \
    .merge(family_status_dict, how='inner', on='family_status_id')

In [69]:
# сделаем сортировку по убыванию значений в столбце с количеством задолженностей  
table_family_status.sort_values(by='amount_debt_delay', ascending=False)

family_status_id  amount_debt_delay          family_status
0                 0                931        женат / замужем
1                 1                388       гражданский брак
4                 4                274  Не женат / не замужем
3                 3                 85              в разводе
2                 2                 63         вдовец / вдова

В абсолютных цифрах разница между просрочкой в зависимости от семейного положения прослеживается. Однако это впечатление обманчиво, поскольку вполне вероятно, что и количество клиентов с таким статусом будет выше, чем остальных.
Проверим эту гипотезу, присоединив столбец с данными об общем числе клиентов соответствующего статуса и посчитаем процентное соотношение просрочек в каждой группе к общему числу клиентов. 

In [70]:
table_family_status_full = df.groupby('family_status_id', as_index=False) \
    .agg({'gender' : 'count'}) \
    .rename(columns={'gender' : 'all_client_family_status'})

In [71]:
# присоединим таблицу с количеством клиентов по каждому статусу
table_family_status = table_family_status.merge(table_family_status_full, how='inner', on='family_status_id')

In [72]:
# выведем получившийся результат
table_family_status

family_status_id  amount_debt_delay          family_status  \
0                 0                931        женат / замужем   
1                 1                388       гражданский брак   
2                 2                 63         вдовец / вдова   
3                 3                 85              в разводе   
4                 4                274  Не женат / не замужем   

   all_client_family_status  
0                     12344  
1                      4162  
2                       959  
3                      1195  
4                      2810

In [73]:
# добавим столбец с процентным соотношением числа клиентов с просрочками к общему числу клиентов по каждому семейному статусу
table_family_status['procent_debt'] = \
(table_family_status['amount_debt_delay'] / table_family_status['all_client_family_status'] * 100).round(2)

In [74]:
# сделаем сортировку по убыванию процентного соотношения и выведем на экран
table_family_status.sort_values(by='procent_debt', ascending=False)

family_status_id  amount_debt_delay          family_status  \
4                 4                274  Не женат / не замужем   
1                 1                388       гражданский брак   
0                 0                931        женат / замужем   
3                 3                 85              в разводе   
2                 2                 63         вдовец / вдова   

   all_client_family_status  procent_debt  
4                      2810          9.75  
1                      4162          9.32  
0                     12344          7.54  
3                      1195          7.11  
2                       959          6.57

#### Вывод по вопросу 2:

В абсолютных цифрах разница между просрочкой в зависимости от семейного положения имелась, но сильно выделяла одну категорию (женат / замужем) по сравнению с остальными.

Однако при переводе соответствующих значений в проценты по сравнению с общим количеством клиентов с соответствующим статусом, выяснилось, что просрочку чаще допускают другие группы клиентов.

Из предоставленных данных следует наличие зависимости между семейным положением и возвратом кредита в срок, например, клиенты со статусами "гражданский брак" и "Не женат / не замужем" в среднем на 2% чаще допускают нарушение обязательств чем лица, состоящие в браке и почти на 3% чаще, чем ранее имевшие брачные отношения (со статусами "вдовец/вдова", "в разводе").

#### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Для проверки данной гипотезы создадим сводную таблицу с данными по соответствующим значениям.

In [75]:
table_total_income = df.pivot_table(index=['total_income_category'], values='debt', aggfunc={'debt' : 'sum'}) \
    .reset_index() \
    .rename(columns={'debt' : 'amount_debt_delay'})

In [76]:
# отсортируем получившийся результат по столбцу с количеством просрочек
table_total_income.sort_values(by='amount_debt_delay', ascending=False)

total_income_category  amount_debt_delay
2                     C               1360
3                     B                356
1                     D                 21
0                     E                  2
4                     A                  2

Присоединим столбец с данными об общем числе клиентов и посчитаем процентное соотношение просрочек в каждой группе к общему числу клиентов.

In [77]:
table_total_income_full = df.groupby('total_income_category', as_index=False) \
    .agg({'gender' : 'count'}) \
    .rename(columns={'gender' : 'all_client_income_status'})

In [78]:
table_total_income = table_total_income.merge(table_total_income_full, how='inner', on='total_income_category')

In [79]:
table_total_income

total_income_category  amount_debt_delay  all_client_income_status
0                     E                  2                        22
1                     D                 21                       350
2                     C               1360                     16032
3                     B                356                      5041
4                     A                  2                        25

In [80]:
# добавим столбец с процентным соотношением числа клиентов с просрочками 
# к общему числу клиентов по соответствующей группе дохода
table_total_income['procent_debt'] = \
(table_total_income['amount_debt_delay'] / table_total_income['all_client_income_status'] * 100).round(2)

In [81]:
# сделаем сортировку по процентному соотношению и выведем на экран
table_total_income.sort_values(by='procent_debt', ascending=False)

total_income_category  amount_debt_delay  all_client_income_status  \
0                     E                  2                        22   
2                     C               1360                     16032   
4                     A                  2                        25   
3                     B                356                      5041   
1                     D                 21                       350   

   procent_debt  
0          9.09  
2          8.48  
4          8.00  
3          7.06  
1          6.00

#### Вывод по вопросу 3:

В абсолютных цифрах разница между просрочкой в зависимости от группы доходов имелась, но сильно выделяла две категории ('D' и 'B') по сравнению с остальными. 

При переводе соответствующих значений в проценты по сравнению с общим количеством клиентов в соответствующей группе, данные изменились. Выяснилось, что реже всего просрочку допускают клиенты с относительно низкими доходами (категория 'D') и лица, с доходами выше среднего (категория 'B').

Чаще всего просрочку допускают лица с самыми низкими доходами (категория 'E').

#### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

Для проверки данной гипотезы создадим сводную таблицу с данными по соответствующим значениям.

In [82]:
table_purpose = df.pivot_table(index=['purpose_category'], values='debt', aggfunc={'debt' : 'sum'}) \
    .reset_index() \
    .rename(columns={'debt' : 'amount_debt_delay'})

In [83]:
# отсортируем получившийся результат по столбцу с количеством просрочек
table_purpose.sort_values(by='amount_debt_delay', ascending=False)

purpose_category  amount_debt_delay
1  операции с недвижимостью                782
0    операции с автомобилем                403
2     получение образования                370
3        проведение свадьбы                186

Присоединим столбец с данными об общем числе клиентов и посчитаем процентное соотношение просрочек в каждой группе к общему числу клиентов.

In [84]:
table_purpose_full = df.groupby('purpose_category', as_index=False) \
    .agg({'gender' : 'count'}) \
    .rename(columns={'gender' : 'all_client_purpose'})

In [85]:
# присоединим таблицу с количеством клиентов по соответствующей группе дохода
table_purpose = table_purpose.merge(table_purpose_full, how='inner', on='purpose_category')

In [86]:
# добавим столбец с процентным соотношением числа клиентов с просрочками 
# к общему числу клиентов по соответствующей группе дохода
table_purpose['procent_debt'] = (table_purpose['amount_debt_delay'] / table_purpose['all_client_purpose'] * 100).round(2)

In [87]:
# сделаем сортировку по процентному соотношению и выведем на экран
table_purpose.sort_values(by='procent_debt', ascending=False)

purpose_category  amount_debt_delay  all_client_purpose  \
0    операции с автомобилем                403                4308   
2     получение образования                370                4014   
3        проведение свадьбы                186                2335   
1  операции с недвижимостью                782               10813   

   procent_debt  
0          9.35  
2          9.22  
3          7.97  
1          7.23

#### Вывод по вопросу 4:

В абсолютных цифрах разница между просрочкой в зависимости от цели кредита имелась, но сильно выделяла категорию клиентов с операциями с недвижимостью по сравнению с остальными. 

При переводе соответствующих значений в проценты по сравнению с общим количеством клиентов в соответствующей группе, выяснилось, что клиенты, которые берут кредит для проведения операций с недвижимостью реже по сравнению с остальными категориями допускают просрочку.

Чаще всего просрочку допускают клиенты, которые берут кредит на операции с автомобилем и получение образования.

## Общий вывод:

Мы проверили 4 гипотезы, проведя соответствующее исследование, в ходе которого установили, что исходя из предоставленных данных, показатели отсутствия просрочки в среднем лучше у лиц обладающих следующими признаками:
- состоящими или ранее состоявшими в брачных отношениях;
- не имеющими на иждивении детей;
- имеющими относительно низкие доходы (от 30 001 до 50 000) или доходы выше среднего (от 200 001 до 1 000 000);
- целью получения кредита у которых является проведение операций с недвижимостью (покупка/ремонт/строительство и т.п.).

Указанные сведения можно принять во внимание при корректировке скорринговой модели принятия решения о выдаче кредита.